### PR TITLE
issue: 749133 IGMPv3 - Source-Specific Multicast - sw implementation

### PR DESCRIPTION
**Pull request by DanielLibenson**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/201/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/215/ for details (Mellanox internal link).

File: src/vma/sock/sockinfo_udp.cpp [hunk 1]
```diff
@@ -382,6 +384,7 @@ sockinfo_udp::sockinfo_udp(int fd) throw (vma_exception) :
 {
 	si_udp_logfunc("");
 
+	m_mc_num_grp_with_src_filter = 0;
```
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.cpp]: please move as part of member initializer list

File: src/vma/sock/sockinfo_udp.cpp [hunk 2]
```diff
@@ -913,114 +916,113 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
...
-							mreq.imr_interface.s_addr = p_ndev->get_local_addr();
-						} else {
-							struct sockaddr_in src_addr;
-							if (get_ipv4_from_ifindex(p_mreqn->imr_ifindex, &src_addr) == 0) {
-								mreq.imr_interface.s_addr = src_addr.sin_addr.s_addr;
-							} else {
-								si_udp_logdbg("setsockopt(%s) will be passed to OS for handling, can't get address of interface index %d ", setsockopt_ip_opt_to_str(__optname), p_mreqn->imr_ifindex);
-								break;
+					struct ip_mreq_source mreqprm = {{((struct ip_mreq_source*)__optval)->imr_multiaddr.s_addr},{((struct ip_mreq_source*)__optval)->imr_interface.s_addr},{0}};
+					if (((IP_ADD_MEMBERSHIP == __optname) || (IP_DROP_MEMBERSHIP == __optname)) && (__optlen >= sizeof(struct ip_mreqn))) {
```
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.cpp]: can you suggest code which use single check for IP_ADD|DROP_MEMBERSHIP so the if() and the above switch-case appear once? maybe use helper function for common code, and parsing under each relevant part.

File: src/vma/sock/sockinfo_udp.cpp [hunk 3]
```diff
@@ -913,114 +916,113 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
...
+					if (!m_sock_offload) {
+						si_udp_logdbg("VMA Rx Offload is Disabled! calling OS setsockopt() for IPPROTO_IP, %s", setsockopt_ip_opt_to_str(__optname));
+						break;
+					}
 
-				if (__optval == NULL || (__optlen < sizeof(struct ip_mreq))) {
-					si_udp_logdbg("IPPROTO_IP, %s; Bad optval! calling OS setsockopt()", setsockopt_ip_opt_to_str(__optname));
-					break;
-				}
+					if ((NULL == __optval) || (__optlen > sizeof(struct ip_mreq_source) || (__optlen < sizeof(struct ip_mreq)))) {
```
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.cpp]: should we also check if smaller the 'struct ip_mreq'? like original code?

File: src/vma/sock/sockinfo_udp.h [hunk 4]
```diff
@@ -74,9 +74,26 @@ namespace std {
 }
 typedef std::tr1::unordered_map<sock_addr, dst_entry*> dst_entry_map_t;
 
-// Multicast Request list
-typedef std::list<struct ip_mreq> ip_mreq_list_t;
-typedef std::tr1::unordered_map<in_addr_t, int> mc_memberships_map_t;
+
+struct mc_pending_pram
+{
+    struct in_addr imr_multiaddr;
```
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.h]: fix struct member alignment

File: src/vma/sock/sockinfo_udp.h [hunk 5]
```diff
@@ -74,9 +74,26 @@ namespace std {
...
+// Multicast pending list
+typedef std::list<struct mc_pending_pram> mc_pram_list_t;
+typedef std::tr1::unordered_map<in_addr_t, std::tr1::unordered_map<in_addr_t, int> > mc_memberships_map_t;
+
+
+typedef enum
+{
+	MCSF_NO_IN_USE = 0x00,
+	MCSF_ADD_DROP,
+	MCSF_BLOCK_UNBLOCK,
```
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.h]: not in use enum, please comment out with explanation..

**Comment by rosenbaumalex** [comment]
please add testing tool.
dedicated or better yet patch to sockperf to allow using the SSM from command line or feedfile

File: src/vma/sock/sockinfo_udp.cpp [hunk 6]
```diff
@@ -913,114 +915,114 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
...
-						} else {
-							struct sockaddr_in src_addr;
-							if (get_ipv4_from_ifindex(p_mreqn->imr_ifindex, &src_addr) == 0) {
-								mreq.imr_interface.s_addr = src_addr.sin_addr.s_addr;
-							} else {
-								si_udp_logdbg("setsockopt(%s) will be passed to OS for handling, can't get address of interface index %d ", setsockopt_ip_opt_to_str(__optname), p_mreqn->imr_ifindex);
-								break;
+					struct ip_mreq_source mreqsrc = {{((struct ip_mreq_source*)__optval)->imr_multiaddr.s_addr},{((struct ip_mreq_source*)__optval)->imr_interface.s_addr},{0}};
+					if (((IP_ADD_MEMBERSHIP == __optname) || (IP_DROP_MEMBERSHIP == __optname)) && (__optlen >= sizeof(struct ip_mreqn))) {
+							struct ip_mreqn* p_mreqn = (struct ip_mreqn*)__optval;
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_udp.cpp]: Not sure about the if logic.
Please consider:
```code
	if (((IP_ADD_MEMBERSHIP == __optname) || (IP_DROP_MEMBERSHIP == __optname)) {
						if (__optlen >= sizeof(struct ip_mreqn)) {
                                                ....
                            }
       else {
       // handle source IP
      }
```

File: src/vma/sock/sockinfo_udp.cpp (hunk 3 above)
Comment (DanielLibenson) [inline src/vma/sock/sockinfo_udp.cpp]: It still there [(__optlen < sizeof(struct ip_mreq)]

File: src/vma/sock/sockinfo_udp.cpp (hunk 2 above)
Comment (DanielLibenson) [inline src/vma/sock/sockinfo_udp.cpp]: I'm not sure if there is a simple way to separate the cases; but there is a bug inside that if() that I will fix.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/212/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/225/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/322/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/340/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/337/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/355/ for details (Mellanox internal link).

**Comment by rosenbaumalex** [comment]
@DanielLibenson please rebase. 
Branch can't be merged with recent master changes.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/351/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/369/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/436/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/452/ for details (Mellanox internal link).

File: src/vma/sock/sockinfo_udp.cpp [hunk 7]
```diff
@@ -1787,6 +1787,17 @@ bool sockinfo_udp::rx_input_cb(mem_buf_desc_t* p_desc, void* pv_fd_ready_array)
...
 	}
 
+	if ((m_mc_num_grp_with_src_filter) && (IN_MULTICAST_N(p_desc->path.rx.dst.sin_addr.s_addr))) {
+		in_addr_t mc_grp = p_desc->path.rx.dst.sin_addr.s_addr;
+		in_addr_t mc_src = p_desc->path.rx.src.sin_addr.s_addr;
+
+		if ((m_mc_memberships_map.find(mc_grp) == m_mc_memberships_map.end()) ||
+			((0 < m_mc_memberships_map[mc_grp].size()) &&
+			(m_mc_memberships_map[mc_grp].find(mc_src) == m_mc_memberships_map[mc_grp].end()))) {
+				return false;
```
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.cpp]: please add 'discard' log message similar to other test cases: `si_udp_logfunc("rx packet discarded - multicast source mismatch ...");`

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/439/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/455/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/443/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/459/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/446/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/462/ for details (Mellanox internal link).

**Comment by rosenbaumalex** [comment]
Merged in https://github.com/Mellanox/libvma/commit/4a6348771d66fb75016c19137982885b5590fda0

**Pull request closed without merging**